lalrpop/src/api/test.rs

```diff
@@ -0,0 +1,215 @@
+use std::env::{current_dir, set_current_dir, set_var, temp_dir};
+use std::fs;
+use std::path;
+use std::sync::{LockResult, Mutex, MutexGuard};
+
+use super::*;
+
+// tests may be run in parallel and these tests affect global state, so lock
+static API_TEST_MUTEX: Mutex<i32> = Mutex::new(0);
+
+const TEST_DIR: &str = "lalrpop-test";
+const CUSTOM_TEST_DIR: &str = "lalrpop-test2";
+
+#[derive(Debug, PartialEq)]
+enum GenFileLoc {
+    Src,
+    Other,
+    Root,
+    OutDir,
+    OutDirSlashOther,
+    CustomOut,
+    DoesntExist,
+}
+
+// This holds state during the test to be cleaned up on drop
+struct TestState {
+    orig_dir: path::PathBuf,
+    lock: LockResult<MutexGuard<'static, i32>>,
```

Comment:
Maybe this is a silly question since i don't work with Mutex's often but why do we hold the LockResult and not the MutexGuard here?

Reply:
No, good question.  The direct answer is "I don't really remember from when I did earlier rounds of this months ago".

But at the end of the day, with the new drop() implementation for teardown, it shouldn't matter much.  We return an error if a different test failed.  That would matter if they didn't have cleaned up state, but they should always clean up state on drop() now.  I added some handling for defense in depth.